### PR TITLE
gh10841 - Purchase Candidate REST implementation

### PR DESCRIPTION
**Pull request by adi-stefan**
https://github.com/metasfresh/metasfresh/issues/10841

### CODE REVIEW

File: backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/createPurchaseCandidate.feature [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      | attributeName | attributeCode | valueStr | valueNumber | valueDate |
+      |               |               |          |             |           |
+    Given the user adds a purchase candidate enqueue request
+      | ExternalHeaderId | ExternalLineId |
+      | H1               | L1             |
+
+    And the purchase candidate request is set in context
+    When the metasfresh REST-API endpoint path 'api/v2-pre/order/purchase/createCandidates' receives a 'POST' request with the payload from context and responds with '200' status code
+
+  Scenario: create a purchase candidate
```
Comment (metas-ts) [inline backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/createPurchaseCandidate.feature:26]:
```suggestion
```

**Review by metas-ts — CHANGES_REQUESTED** [review]

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/document/impl/DocumentLocationBL.java [hunk 2]
```diff
@@ -124,7 +129,7 @@ public void setDeliveryToAddress(final IDocumentDeliveryLocation docDeliveryLoca
 				return;
 			}
 
-			final I_M_Warehouse warehouse = docDeliveryLocation.getM_Warehouse();
+			final I_M_Warehouse warehouse = Services.get(IWarehouseDAO.class).getById(WarehouseId.ofRepoId(warehouseId));
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/document/impl/DocumentLocationBL.java]: Extracted

File: backend/de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/warehouse/api/impl/WarehouseBL.java [hunk 3]
```diff
@@ -133,7 +135,7 @@ private I_C_Location getC_Location(final I_M_Warehouse warehouse)
 	{
 		Check.assumeNotNull(warehouse, "warehouse not null");
 
-		final I_C_BPartner_Location bpLocation = warehouse.getC_BPartner_Location();
+		final I_C_BPartner_Location bpLocation = Services.get(IBPartnerDAO.class).getBPartnerLocationById(BPartnerLocationId.ofRepoIdOrNull(warehouse.getC_BPartner_ID(),warehouse.getC_BPartner_Location_ID()));
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/warehouse/api/impl/WarehouseBL.java]: extracted

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/order/impl/CreatePurchaseCandidatesService.java [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
+package de.metas.rest_api.order.impl;
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/order/impl/CreatePurchaseCandidatesService.java]: moved.

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/order/impl/CreatePurchaseCandidatesService.java [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+			throw new MissingPropertyException("price", request);
+		}
+
+		final PurchaseOrderPricingInfo pricingInfo = PurchaseOrderPricingInfo.builder()
+				.productId(productId)
+				.orgId(orgId)
+				.quantity(quantity)
+				.bpartnerId(vendorId)
+				.datePromised(datePromised)
+				.discount(discount)
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/order/impl/CreatePurchaseCandidatesService.java]: you're right, removed.

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/order/impl/CreatePurchaseCandidatesService.java [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+	private IPricingResult getPriceAndDiscount(final PurchaseOrderPricingInfo pricingInfo)
+	{
+		return PurchaseOrderPriceCalculator.builder()
+				.pricingBL(pricingBL)
+				.pricingInfo(pricingInfo)
+				.build()
+				.calculatePrice();
+	}
+
+	private ZonedDateTime getOrDefaultDatePromised(final @Nullable ZonedDateTime purchaseDatePromised, final OrgId orgId)
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/order/impl/CreatePurchaseCandidatesService.java]: as discussed, keeping it as it is.

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/order/impl/CreatePurchaseCandidatesService.java [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+	{
+		return purchaseDatePromised != null ?
+				purchaseDatePromised :
+				SystemTime.asZonedDateTime(orgDAO.getTimeZone(orgId));
+	}
+
+	@NonNull
+	private Quantity getQuantity(final JsonPurchaseCandidateCreateItem request)
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/order/impl/CreatePurchaseCandidatesService.java]: moved.

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/order/impl/CreatePurchaseCandidatesService.java [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+		{
+			handleIncorrectProductIdentifier(productIdentifier);
+		}
+
+		return result;
+	}
+
+	private void handleIncorrectProductIdentifier(final String productIdentifier)
+	{
+		throw MissingResourceException.builder()
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/order/impl/CreatePurchaseCandidatesService.java]: changed exception class.

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/order/impl/CreatePurchaseCandidatesService.java [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+		{
+			handleIncorrectWarehouseIdentifier(warehouseIdentifier);
+		}
+
+		return result;
+	}
+
+	private void handleIncorrectWarehouseIdentifier(final String warehouseIdentifier)
+	{
+		throw MissingResourceException.builder()
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/order/impl/CreatePurchaseCandidatesService.java]: changed exception class.

File: backend/de.metas.business/src/main/java/de/metas/uom/IUOMDAO.java [hunk 10]
```diff
@@ -78,4 +79,6 @@
 
 	@NonNull
 	UOMType getUOMTypeById(UomId uomId);
+
+	I_C_UOM getByProductID(ProductId productId);
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.business/src/main/java/de/metas/uom/IUOMDAO.java]: dropped like it's hot.

File: backend/de.metas.purchasecandidate.base/src/main/java/de/metas/purchasecandidate/PurchaseCandidateSource.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+import de.metas.util.lang.ReferenceListAwareEnum;
+import lombok.NonNull;
+import org.adempiere.exceptions.AdempiereException;
+
+import javax.annotation.Nullable;
+import java.util.Arrays;
+
+public enum PurchaseCandidateSource implements ReferenceListAwareEnum
+{
+	MaterialDisposition("MD"),
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.purchasecandidate.base/src/main/java/de/metas/purchasecandidate/PurchaseCandidateSource.java]: changed to use constants

File: backend/de.metas.purchasecandidate.base/src/main/java/de/metas/purchasecandidate/purchaseordercreation/PurchaseOrderPricingInfo.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+public class PurchaseOrderPricingInfo
+{
+	@NonNull OrgId orgId;
+	@Nullable ZoneId orgTimeZone;
+	@NonNull BPartnerId bpartnerId;
+	@NonNull ZonedDateTime datePromised;
+	@NonNull ProductId productId;
+	@Nullable CountryId countryId;
+	@NonNull Quantity quantity;
+	@NonNull BigDecimal discount;
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.purchasecandidate.base/src/main/java/de/metas/purchasecandidate/purchaseordercreation/PurchaseOrderPricingInfo.java]: removed.

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/modelvalidator/M_Warehouse.java [hunk 13]
```diff
@@ -48,7 +53,7 @@ public void init()
 	@CalloutMethod(columnNames = I_M_Warehouse.COLUMNNAME_C_BPartner_Location_ID)
 	public void syncLocation(final I_M_Warehouse warehouse)
 	{
-		final I_C_BPartner_Location bpLocation = warehouse.getC_BPartner_Location();
+		final I_C_BPartner_Location bpLocation = bpartnerDAO.getBPartnerLocationById(BPartnerLocationId.ofRepoIdOrNull(warehouse.getC_BPartner_ID(),warehouse.getC_BPartner_Location_ID()));
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/modelvalidator/M_Warehouse.java]: refactored to account for potential null locationId.

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/sales/purchasePlanning/view/PurchaseRowsSaver.java [hunk 14]
```diff
@@ -248,11 +249,12 @@ else if (!candidatesToUpdate.isEmpty())
 					.profitInfoOrNull(profitInfo)
 					.forecastLineId(candidatesGroup.getForecastLineId())
 					.dimension(candidatesGroup.getDimension())
+					.source(PurchaseCandidateSource.SalesOrder)
 					//
 					.build();
 
 			candidatesChanged.add(newCandidate);
-			qtyToPurchaseRemainingOfGroup = qtyToPurchaseRemainingOfGroup.toZero();
+			qtyToPurchaseRemainingOfGroup.toZero();
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/order/sales/purchasePlanning/view/PurchaseRowsSaver.java]: yeh, cleaned up.

File: misc/de-metas-common/de-metas-common-rest_api/src/main/java/de/metas/rest_api/order/JsonWorkPackage.java [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
+package de.metas.rest_api.order;
+
+public class JsonWorkPackage
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply misc/de-metas-common/de-metas-common-rest_api/src/main/java/de/metas/rest_api/order/JsonWorkPackage.java]: dropped.

File: backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/createPurchaseCandidate.feature (hunk 1 above)
Comment (adi-stefan) [reply backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/createPurchaseCandidate.feature]: removed.

File: backend/de.metas.business/src/main/java/de/metas/uom/IUOMDAO.java [hunk 16]
```diff
@@ -23,6 +23,7 @@
 package de.metas.uom;
 
 import de.metas.i18n.ITranslatableString;
+import de.metas.product.ProductId;
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/uom/IUOMDAO.java:26]: ok, now pls also remove this little critter

File: backend/de.metas.business/src/main/java/de/metas/uom/impl/UOMDAO.java [hunk 17]
```diff
@@ -24,6 +24,7 @@
 
 import de.metas.cache.CCache;
 import de.metas.i18n.ITranslatableString;
+import de.metas.product.ProductId;
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/uom/impl/UOMDAO.java:27]: here as well, pls clean up the unused imports

**Review by metas-ts — APPROVED** [review]

**Pull request merged**